File: ios/MoodleLMS/Features/Grades/GradesRepository.swift
import Foundation

struct GradeItem: Identifiable, Hashable {
    let id: Int
    let name: String
    let grade: String?
    let percentage: String?
}

final class GradesRepository {
    private let client: MoodleClient
    private let tokenStore: KeychainStore

    init(client: MoodleClient, tokenStore: KeychainStore) {
        self.client = client
        self.tokenStore = tokenStore
    }

    func loadGrades(courseId: Int) async throws -> [GradeItem] {
        guard let userId = tokenStore.userId() else { throw MoodleError.unauthenticated }
        let response: GradeReportResponse = try await client.callFunction(
            "gradereport_user_get_grade_items",
            params: ["userid": String(userId), "courseid": String(courseId)]
        )
        return (response.usergrades.first?.gradeitems ?? []).map { dto -> GradeItem in
            let rawName = (dto.itemname?.isEmpty == false ? dto.itemname! :
                            (dto.itemtype?.lowercased() == "course" ? "Course total" : "Grade item"))
            return GradeItem(
                id: dto.id,
                name: rawName.htmlUnescaped,
                grade: clean(dto.gradeformatted),
                percentage: clean(dto.percentageformatted)
            )
        }
    }

    /// Moodle returns "-" or "- %" for ungraded items. Display nothing instead.
    private func clean(_ value: String?) -> String? {
        guard let trimmed = value?.trimmingCharacters(in: .whitespaces), !trimmed.isEmpty else { return nil }
        if trimmed == "-" { return nil }
        let stripped = trimmed.replacingOccurrences(of: "[\\s%-]", with: "", options: .regularExpression)
        return stripped.isEmpty ? nil : trimmed
    }
}

File: android/app/src/main/java/io/itcorner/moodle/feature/grades/GradesRepository.kt
package io.itcorner.moodle.feature.grades

import io.itcorner.moodle.core.auth.TokenStore
import io.itcorner.moodle.core.network.MoodleApi
import io.itcorner.moodle.core.network.MoodleApiException
import io.itcorner.moodle.feature.courses.htmlUnescape
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

data class GradeItem(
    val id: Long,
    val name: String,
    val grade: String?,
    val percentage: String?,
)

@Singleton
class GradesRepository @Inject constructor(
    private val api: MoodleApi,
    private val tokenStore: TokenStore,
) {
    suspend fun loadGrades(courseId: Long): Result<List<GradeItem>> {
        val userId = tokenStore.userId() ?: return Result.failure(IllegalStateException("Not signed in"))
        return try {
            val response = api.getGradeItems(userId, courseId)
            val items = response.usergrades.firstOrNull()?.gradeitems.orEmpty().map {
                val name = (it.itemname?.takeIf { v -> v.isNotBlank() }
                    ?: if (it.itemtype.equals("course", ignoreCase = true)) "Course total" else "Grade item")
                    .htmlUnescape()
                GradeItem(
                    id = it.id,
                    name = name,
                    grade = it.gradeformatted.cleaned(),
                    percentage = it.percentageformatted.cleaned(),
                )
            }
            Result.success(items)
        } catch (e: IOException) {
            Result.failure(e)
        } catch (e: MoodleApiException) {
            Result.failure(e)
        }
    }

    private fun String?.cleaned(): String? {
        if (this.isNullOrBlank()) return null
        val trimmed = trim()
        // Moodle uses "-" or "- %" for ungraded items — show nothing instead.
        return when {
            trimmed == "-" -> null
            trimmed.replace(Regex("[\\s%-]"), "").isEmpty() -> null
            else -> trimmed
        }
    }
}
